Adopt work interval API

pvollan opened the pull request:
#### f31d0edc7b5e836fb7570ba0b4a2e28a0d092de2
<pre>
Adopt work interval API
<a href="https://bugs.webkit.org/show_bug.cgi?id=246829">https://bugs.webkit.org/show_bug.cgi?id=246829</a>
rdar://101398504

Reviewed by NOBODY (OOPS!).

Adopt work interval API on macOS and iOS.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/cocoa/WorkgroupInterval.cpp: Added.
(WTF::WorkgroupInterval::WorkgroupInterval):
(WTF::WorkgroupInterval::createInterval):
(WTF::WorkgroupInterval::join):
(WTF::WorkgroupInterval::leave):
(WTF::WorkgroupInterval::start):
(WTF::WorkgroupInterval::update):
(WTF::WorkgroupInterval::finish):
* Source/WTF/wtf/cocoa/WorkgroupInterval.h: Added.
(WTF::WorkgroupIntervalScope::WorkgroupIntervalScope):
(WTF::WorkgroupIntervalScope::~WorkgroupIntervalScope):
* Source/WebCore/page/Page.cpp:
(WebCore::workgroupRendering):
(WebCore::Page::~Page):
(WebCore::Page::updateRendering):
(WebCore::Page::setIsVisuallyIdleInternal):
* Source/WebCore/page/Page.h:
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f31d0edc7b5e836fb7570ba0b4a2e28a0d092de2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93808 "Passed style check") | [❌ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3003 "Hash f31d0edc for PR 5602 does not build (failure)") | [❌ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/24390 "Hash f31d0edc for PR 5602 does not build (failure)") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103446 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/163771 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97801 "Passed tests") | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3019 "Hash f31d0edc for PR 5602 does not build (failure)") | [❌ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31246 "Hash f31d0edc for PR 5602 does not build (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86136 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/99457 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99470 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/76/builds/3019 "Hash f31d0edc for PR 5602 does not build (failure)") | [❌ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80240 "Hash f31d0edc for PR 5602 does not build (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29173 "Passed tests") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/76/builds/3019 "Hash f31d0edc for PR 5602 does not build (failure)") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/43/builds/24390 "Hash f31d0edc for PR 5602 does not build (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72128 "Passed tests") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/85063 "Hash f31d0edc for PR 5602 does not build (failure)") | [❌ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37641 "Hash f31d0edc for PR 5602 does not build (failure)") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/43/builds/24390 "Hash f31d0edc for PR 5602 does not build (failure)") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/80193 "Built successfully") | 
| | [❌ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35499 "Hash f31d0edc for PR 5602 does not build (failure)") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/43/builds/24390 "Hash f31d0edc for PR 5602 does not build (failure)") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/27663 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39377 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/61/builds/80240 "Hash f31d0edc for PR 5602 does not build (failure)") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/82837 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41311 "Built successfully") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/43/builds/24390 "Hash f31d0edc for PR 5602 does not build (failure)") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/18721 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebCore/page/Page.cpp:
```diff
@@ -435,6 +443,11 @@ Page::~Page()
+#if PLATFORM(COCOA)
+    if (m_intervalStarted)
+        workgroupRendering().finish();
+#endif
```
Because the workgroup interval is scoped to Page::updateRendering, it's impossible in a well-formed program for ~Page() to run while we have a live interval -- that would mean that we had deleted 'this' in the midst of a member function. So, I think you can remove this.

geoffreygaren commented on Source/WebCore/page/Page.cpp:
```diff
@@ -2064,6 +2086,13 @@ void Page::setIsVisuallyIdleInternal(bool isVisuallyIdle)
+        m_intervalStarted = false;
+    }
+#endif
+
```
Because the workgroup interval is scoped to Page::updateRendering, it's impossible in a well-formed program for a visually idle notification to be processed while we have a live interval -- since notifications are only processed when we return to the run loop. So, I think you can remove this.

geoffreygaren commented on Source/WebCore/page/Page.h:
```diff
@@ -1348,6 +1352,11 @@ class Page : public Supplementable<Page>, public CanMakeWeakPtr<Page> {
+    
+#if PLATFORM(COCOA)
+    bool m_joined { false };
+    bool m_intervalStarted { false };
```
If I'm right above, then you can remove m_intervalStarted.

geoffreygaren commented on Source/WebCore/page/Page.h:
```diff
@@ -1348,6 +1352,11 @@ class Page : public Supplementable<Page>, public CanMakeWeakPtr<Page> {
     ContentSecurityPolicyModeForExtension m_contentSecurityPolicyModeForExtension { ContentSecurityPolicyModeForExtension::None };
+    
+#if PLATFORM(COCOA)
+    bool m_joined { false };
```
I don't think you need this because you're using call_once instead.

pvollan replied:
Fixed!

geoffreygaren commented on Source/WTF/wtf/cocoa/WorkgroupInterval.cpp:
```diff
@@ -0,0 +1,122 @@
+{
+    Locker locker { m_lock };
+    if (!m_instance)
+        m_instance = work_interval_instance_alloc(m_interval);
```
Should we just allocate this in our constructor?

pvollan replied:
Fixed!

geoffreygaren commented on Source/WTF/wtf/cocoa/WorkgroupInterval.cpp:
```diff
@@ -0,0 +1,122 @@
+namespace WTF {
+
+WorkgroupInterval::WorkgroupInterval(const char* name)
+    : m_interval(createInterval(name))
```
Should we deallocate this in our destructor?

geoffreygaren commented on Source/WTF/wtf/cocoa/WorkgroupInterval.h:
```diff
@@ -0,0 +1,73 @@
+    work_interval_t m_interval { nullptr };
+    work_interval_instance_t m_instance { nullptr };
+    std::optional<std::thread::id> m_startThread;
+    Lock m_lock;
```
I don't think we're using the multi-threading feature anymore.

pvollan replied:
Thanks! The latest patch should address these issues.

smfr commented on Source/WTF/wtf/cocoa/WorkgroupInterval.cpp:
```diff
@@ -0,0 +1,126 @@
+    const int rv = work_interval_create(&workIntervalHandle, WORK_INTERVAL_TYPE_CA_CLIENT | WORK_INTERVAL_FLAG_JOINABLE);
+    RELEASE_ASSERT(!rv);
+    RELEASE_ASSERT(workIntervalHandle);
+    WTFLogAlways("Workgroup interval = %p", workIntervalHandle);
```
Need to remove the logging.

smfr commented on Source/WTF/wtf/cocoa/WorkgroupInterval.cpp:
```diff
@@ -0,0 +1,126 @@
+    return !rv;
+}
+
+bool WorkgroupInterval::start(uint64_t startOffsetNanoSeconds, uint64_t deadlineNanoSeconds)
```
I think this should just take a `Seconds` duration. We always call it to start "now".

smfr commented on Source/WTF/wtf/cocoa/WorkgroupInterval.cpp:
```diff
@@ -0,0 +1,126 @@
+    if (m_startThread)
+        return false;
+    mach_timebase_info_data_t timebase;
+    mach_timebase_info(&timebase);
```
Would be nicer to use `MonotonicTime::toMachAbsoluteTime` math, rather than replicating that math here.

smfr commented on Source/WTF/wtf/cocoa/WorkgroupInterval.cpp:
```diff
@@ -0,0 +1,126 @@
+{
+    mach_timebase_info_data_t timebase;
+    mach_timebase_info(&timebase);
+    int64_t deadlineAbsolute = (deadlineNanoSeconds * timebase.denom) / timebase.numer;
```
Ditto.

smfr commented on Source/WTF/wtf/cocoa/WorkgroupInterval.cpp:
```diff
@@ -0,0 +1,126 @@
+        return false;
+    }
+    if (*m_startThread != std::this_thread::get_id()) {
+        WTFLogAlways("Workgroup finish error, start called on different thread");
```
Seems like this should be an ASSERT

pvollan replied:
Will fix!

smfr commented on Source/WTF/wtf/cocoa/WorkgroupInterval.h:
```diff
@@ -0,0 +1,76 @@
+    WTF_EXPORT_PRIVATE ~WorkgroupInterval();
+    WTF_EXPORT_PRIVATE bool join();
+    WTF_EXPORT_PRIVATE bool leave();
+    WTF_EXPORT_PRIVATE bool start(uint64_t startOffsetNanoSeconds, uint64_t deadlineNanoSeconds);
```
Use Seconds here.

pvollan replied:
Will fix!

smfr commented on Source/WTF/wtf/cocoa/WorkgroupInterval.h:
```diff
@@ -0,0 +1,76 @@
+class WorkgroupIntervalScope {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    WorkgroupIntervalScope(WorkgroupInterval& interval, int64_t startOffsetNanoSeconds, uint64_t deadlineNanoSeconds)
```
Use Seconds here.

pvollan replied:
Will fix.

smfr commented on Source/WebCore/page/Page.cpp:
```diff
@@ -1675,6 +1683,15 @@ void Page::updateRendering()
     m_lastRenderingUpdateTimestamp = MonotonicTime::now();
 
+#if HAVE(WORK_INTERVAL_API)
+    static std::once_flag onceFlag;
```
This code needs to move somewhere else. The `WorkgroupIntervalScope` should be a one-liner here.

Also consider making a stub for other platforms so that there is no #ifdeffing needed here.

pvollan replied:
Thanks for reviewing! I will address these issues.

pvollan committed:
Adopt work interval API
https://bugs.webkit.org/show_bug.cgi?id=246829
rdar://101398504

Reviewed by NOBODY (OOPS!).

Adopt work interval API on macOS and iOS.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/cocoa/WorkgroupInterval.cpp: Added.
(WTF::WorkgroupInterval::WorkgroupInterval):
(WTF::WorkgroupInterval::createInterval):
(WTF::WorkgroupInterval::join):
(WTF::WorkgroupInterval::leave):
(WTF::WorkgroupInterval::start):
(WTF::WorkgroupInterval::update):
(WTF::WorkgroupInterval::finish):
* Source/WTF/wtf/cocoa/WorkgroupInterval.h: Added.
(WTF::WorkgroupIntervalScope::WorkgroupIntervalScope):
(WTF::WorkgroupIntervalScope::~WorkgroupIntervalScope):
* Source/WebCore/page/Page.cpp:
(WebCore::workgroupRendering):
(WebCore::Page::~Page):
(WebCore::Page::updateRendering):
(WebCore::Page::setIsVisuallyIdleInternal):
* Source/WebCore/page/Page.h:
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:

smfr commented on Source/WebCore/page/Page.h:
```diff
@@ -82,6 +82,10 @@
 #endif
 
+#if PLATFORM(COCOA)
+#include <wtf/cocoa/WorkgroupInterval.h>
```
Why is this needed in Page.h?

pvollan replied:
Good catch, this should go in the cpp file. Will fix, this will likely be in a new PR, though.

Thanks for reviewing!

smfr commented on Source/WebCore/page/Page.cpp:
```diff
@@ -262,6 +262,14 @@ static constexpr OptionSet<ActivityState::Flag> pageInitialActivityState()
 }
 
+#if HAVE(WORK_INTERVAL_API)
+static WTF::WorkgroupInterval& workgroupRendering()
```
I'd still prefer to see this outside of Page.cpp.

pvollan replied:
Yes, I agrees. Will fix, likely in a new PR.

Thanks!
